src/random.rs

```diff
@@ -0,0 +1,325 @@
+//! Cryptographically secure random number generation.
+#![warn(missing_docs)]
+
+use core::convert::TryFrom;
+use core::fmt;
+use core::ptr;
+
+use winapi::shared::bcrypt::*;
+use winapi::shared::ntdef::ULONG;
+
+use crate::helpers::{AlgoHandle, Handle};
+use crate::Error;
+
+/// An RNG algorithm provider. Main type that is capable of generating random
+/// numbers.
+pub enum RandomAlgorithm {
+    /// System-preferred algorithm provider.
+    SystemPreferred,
+    /// An already opened provider for a specified algorithm.
+    Specified(RandomAlgoHandle),
+}
+
+/// Wrapper around `AlgoHandle` that can only specify RNG algorithms.
+pub struct RandomAlgoHandle(AlgoHandle);
+
+/// Kind of a random algorithm to be used when generating numbers.
+#[derive(Clone, Copy)]
+pub enum RandomAlgorithmKind {
+    /// Use the system-preferred random number generator algorithm.
+    /// Implies setting `BCRYPT_USE_SYSTEM_PREFERRED_RNG` flag when calling
+    /// `BCryptGenRandom`.
+    ///
+    /// This is only supported at `PASSIVE_LEVEL` IRQL. For more information,
+    /// see [`Remarks`](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom#remarks).
+    /// *Windows Vista*:  This flag is not supported.
+    SystemPreferred,
+    /// Use the specified random number generator algorithm.
+    Specified(RandomAlgorithmId),
+}
+
+/// Random number generation algorithms supported RNG algorithms by Windows CNG API.
+#[derive(Clone, Copy)]
+pub enum RandomAlgorithmId {
+    /// The random-number generator algorithm.
+    /// Standard: FIPS 186-2, FIPS 140-2, NIST SP 800-90
+    ///
+    /// Beginning with Windows Vista with SP1 and Windows Server 2008, the
+    /// random number generator is based on the AES counter mode specified in
+    /// the NIST SP 800-90 standard.
+    ///
+    /// *Windows Vista*: The random number generator is based on the hash-based
+    /// random number generator specified in the FIPS 186-2 standard.
+    ///
+    /// *Windows 8*: Beginning with Windows 8, the RNG algorithm supports
+    /// FIPS 186-3. Keys less than or equal to 1024 bits adhere to FIPS 186-2
+    /// and keys greater than 1024 to FIPS 186-3.
+    RNG,
```

Comment:
`rustc` does the case linting and I wanted to retain the low level naming because it's not so intuitive (it's weird having a random generation algorithm referred to as 'RNG' where there are other options :smile: ) but you're right, I'll change that.